File: clients/apple/Shared/Model/MeetingTelemetry.swift
import Foundation

enum MeetingTelemetryCategory: String, Equatable {
    case connectionLifecycle = "connection_lifecycle"
    case fallbackLifecycle = "fallback_lifecycle"
    case policyFailure = "policy_failure"
}

struct MeetingTelemetryEvent: Equatable {
    var category: MeetingTelemetryCategory
    var name: String
    var timestamp: Date
    var attributes: [String: String]

    init(
        category: MeetingTelemetryCategory,
        name: String,
        timestamp: Date = Date(),
        attributes: [String: String] = [:]
    ) {
        self.category = category
        self.name = name
        self.timestamp = timestamp
        self.attributes = attributes
    }
}

protocol MeetingTelemetrySink {
    func record(_ event: MeetingTelemetryEvent)
}

struct NoOpMeetingTelemetrySink: MeetingTelemetrySink {
    func record(_ event: MeetingTelemetryEvent) {}
}

File: clients/android/app/src/main/java/io/sora/kaigi/android/MeetingTelemetry.kt
package io.sora.kaigi.android

enum class MeetingTelemetryCategory(val wire: String) {
    ConnectionLifecycle("connection_lifecycle"),
    FallbackLifecycle("fallback_lifecycle"),
    PolicyFailure("policy_failure")
}

data class MeetingTelemetryEvent(
    val category: MeetingTelemetryCategory,
    val name: String,
    val atMs: Long = System.currentTimeMillis(),
    val attributes: Map<String, String> = emptyMap()
)

fun interface MeetingTelemetrySink {
    fun record(event: MeetingTelemetryEvent)
}

object NoOpMeetingTelemetrySink : MeetingTelemetrySink {
    override fun record(event: MeetingTelemetryEvent) = Unit
}
